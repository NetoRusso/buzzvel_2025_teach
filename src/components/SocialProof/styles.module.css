.container {
  width: 100%;
  height: auto;
  min-height: 60vh;
  display: flex;
  align-items: center;
  justify-content: flex-end
}

.content {
  width: 100%;
  max-width: 1440px;
  height: auto;
  min-height: 480px;
  max-height: 640px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: flex-end;
  gap: 80px;
  padding: 32px 16px 16px 16px;
}

.header {
  width: 90%;
  height: auto;
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  justify-content: space-between;
  gap: 48px;
}

.header_title {
  font-size: clamp(1.5rem, 0.7493rem + 3.0564vw, 3.5rem);
  line-height: 110%;
  font-weight: 800;
  color: var(--blue-Gray-900);
}

.header_btns {
  width: 120px;
  height: 48px;
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  justify-content: center;
  gap: 24px;
}

.header_btn {
  transition: all 0.3s ease-in-out;
  cursor: pointer;
}

.header_btn:hover {
  scale: 0.95;
  opacity: 0.9;
}

.cards_box {
  width: 100%;
  max-width: 1440px;
  height: 450px;
  align-self: flex-end;
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  justify-content: flex-start;
  gap: 24px;
  padding: 0 48px;
  cursor: grab;
  overflow-x: auto;
  scroll-snap-type: x mandatory;
  -webkit-overflow-scrolling: touch;
  scrollbar-width: none;
  -ms-overflow-style: none;
}

.cards_box::-webkit-scrollbar {
  display: none;
}

@media screen and (max-width: 840px) {
  .cards_box {
    padding: 0 16px;
  }
}

.card {
  width: 100%;
  min-width: clamp(20rem, 17.654rem + 9.5511vw, 26.25rem);
  height: 375px;
  display: flex;
  flex-flow: column nowrap;
  align-items: flex-start;
  justify-content: space-between;
  padding: 32px;
  border-radius: 16px;
  box-shadow: 0px 0px 10px 0px #00000012, 0px 10px 15px -3px #0000001A;
  background-color: var(--background-white);

}

.card_text {
  font-size: 18px;
  line-height: 160%;
  font-weight: 400;
  color: var(--black);
  user-select: none;
}

.card_ass {
  width: 100%;
  height: 80px;
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  justify-content: left;
  gap: 16px;
  user-select: none;
}

.card_info {
  display: flex;
  flex-flow: column nowrap;
  align-items: flex-start;
  justify-content: center;
}

.card_name {
  font-size: 18px;
  font-weight: 400;
  line-height: 160%;
  color: var(--black);
  user-select: none;
}

.card_role {
  font-size: 16px;
  font-weight: 400;
  line-height: 140%;
  color: var(--blue-Gray-600);
  user-select: none;
}