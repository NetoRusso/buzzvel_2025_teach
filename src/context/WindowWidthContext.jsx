'use client'
import { createContext, useState, useEffect } from "react";

const WindowWidthContext = createContext();

export const WindowWidthProvider = ({ children }) => {
  const [windowWidth, setWindowWidth] = useState(0);
  const [isMobile, setIsMobile] = useState(false);

  useEffect(() => {
    const handleResize = () => {
      setWindowWidth(window.innerWidth);
    }

    window.addEventListener("resize", handleResize);

    handleResize();

    return () => {
      window.removeEventListener("resize", handleResize);
    };
  }, []);

  useEffect(() => {
    if(windowWidth < 840) {
      setIsMobile(true);

    } else if(windowWidth > 840) {
      setIsMobile(false);

    }
  }, [windowWidth]);

  return (
    <WindowWidthContext.Provider value={{ windowWidth, isMobile }}>
      {children}
    </WindowWidthContext.Provider>
  );
};

export { WindowWidthContext };