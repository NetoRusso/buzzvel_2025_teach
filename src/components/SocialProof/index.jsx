'use client'

import { useContext, useRef, useState } from "react";
import Image from "next/image";
import styles from "./styles.module.css";
import { WindowWidthContext } from "@/context/WindowWidthContext";

import prev from "@/assets/SVG/prev_btn.svg";
import next from "@/assets/SVG/next_btn.svg";

import user01 from "@/assets/user_01.png";
import user02 from "@/assets/user_02.png";
import user03 from "@/assets/user_03.png";
import user04 from "@/assets/user_04.png";
import user05 from "@/assets/user_05.png";

const SocialProof = () => {
  const { isMobile } = useContext(WindowWidthContext);
  const cardsContainerRef = useRef(null);

  const userCard = [
    {
      text: "Lacus vestibulum ultricies mi risus, duis non, volutpat nullam non. Magna congue nisi maecenas elit aliquet eu sed consectetur.",
      img: user01,
      name: "Hellen Jummy",
      role: "Financial Counselor"
    },
    {
      text: "Odio rhoncus ornare ut quam. Molestie vel duis quis scelerisque ut id. In tortor turpis viverra sagittis ultrices nisi.",
      img: user02,
      name: "Ralph Edwards",
      role: "Math Teacher"
    },
    {
      text: "Sagittis nunc egestas leo et malesuada urna risus. Morbi proin et cras aliquam. Diam tellus, amet, hac imperdiet.",
      img: user03,
      name: "Hellena John",
      role: "Psychology Student"
    },
    {
      text: "Sapien, sollicitudin et vitae id et laoreet sapien consectetur. Felis egestas egestas amet aliquam sit euismod.",
      img: user04,
      name: "David Oshodi",
      role: "Manager"
    },
    {
      text: "Sapien, sollicitudin et vitae id et laoreet sapien consectetur. Felis egestas egestas amet aliquam sit euismod.",
      img: user05,
      name: "John Doe",
      role: "Engineer"
    }
  ];

  const handleScroll = (direction) => {
    if (cardsContainerRef.current) {
      const scrollAmount = 300;
      const container = cardsContainerRef.current;
      container.scrollBy({ left: direction === "left" ? -scrollAmount : scrollAmount, behavior: "smooth" });
    }
  };

  const [isDragging, setIsDragging] = useState(false);
  const [startX, setStartX] = useState(0);
  const [scrollLeft, setScrollLeft] = useState(0);

  const handleMouseDown = (e) => {
    if (!cardsContainerRef.current) return;
    setIsDragging(true);
    setStartX(e.pageX - cardsContainerRef.current.offsetLeft);
    setScrollLeft(cardsContainerRef.current.scrollLeft);
  };

  const handleMouseLeave = () => setIsDragging(false);
  const handleMouseUp = () => setIsDragging(false);

  const handleMouseMove = (e) => {
    if (!isDragging || !cardsContainerRef.current) return;
    e.preventDefault();
    const x = e.pageX - cardsContainerRef.current.offsetLeft;
    const walk = (x - startX) * 1.2;
    cardsContainerRef.current.scrollLeft = scrollLeft - walk;
  };

  return (
    <section className={styles.container} aria-label="Social proof">
      <div className={styles.content}>
        <div className={styles.header}>
          <h2 className={styles.header_title}>What everyone says</h2>
          {!isMobile && (
            <div className={styles.header_btns}>
              <Image
                src={prev}
                alt="prev"
                width={48}
                height={48}
                loading="lazy"
                className={styles.header_btn}
                onClick={() => handleScroll("left")}
              />
              <Image
                src={next}
                alt="next"
                width={48}
                height={48}
                loading="lazy"
                className={styles.header_btn}
                onClick={() => handleScroll("right")}
              />
            </div>
          )}
        </div>

        <div
          className={styles.cards_box}
          ref={cardsContainerRef}
          onMouseDown={handleMouseDown}
          onMouseLeave={handleMouseLeave}
          onMouseUp={handleMouseUp}
          onMouseMove={handleMouseMove}
        >
          {userCard.map((user, index) => (
            <div className={styles.card} key={index}>
              <p className={styles.card_text}>{user.text}</p>
              <div className={styles.card_ass}>
                <Image
                  src={user.img}
                  alt={`imagem de ${user.name}`}
                  width={64}
                  height={64}
                  loading="lazy"
                  className={styles.card_img}
                />
                <div className={styles.card_info}>
                  <p className={styles.card_name}>{user.name}</p>
                  <p className={styles.card_role}>{user.role}</p>
                </div>
              </div>
            </div>
          ))}
        </div>
      </div>
    </section>
  );
};

export default SocialProof;
