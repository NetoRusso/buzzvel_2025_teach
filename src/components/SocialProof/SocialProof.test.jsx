import React from 'react';
import { render, screen, fireEvent, within } from '@testing-library/react';
import '@testing-library/jest-dom';
import SocialProof from './index';
import { WindowWidthContext } from '@/context/WindowWidthContext';

jest.mock('next/image', () => {
  const ImageComponent = (props) => {
    let { src: srcProp, alt, ...rest } = props;
    let finalSrc = srcProp;
    if (typeof srcProp === 'object' && srcProp !== null && srcProp.src) {
      finalSrc = srcProp.src;
    }
    return <img src={finalSrc} alt={alt || ''} {...rest} />;
  };
  ImageComponent.displayName = 'NextImageMock';
  return {
    __esModule: true,
    default: ImageComponent,
  };
});

const mockWindowWidthContext = (isMobileValue) => ({
  isMobile: isMobileValue,
  windowWidth: isMobileValue ? 400 : 1024,
  isResizing: false,
});

const renderWithContext = (ui, providerProps) => {
  return render(
    <WindowWidthContext.Provider value={providerProps}>
      {ui}
    </WindowWidthContext.Provider>
  );
};

const componentUserCardData = [
    { text: "Lacus vestibulum ultricies mi risus, duis non, volutpat nullam non. Magna congue nisi maecenas elit aliquet eu sed consectetur.", name: "Hellen Jummy", role: "Financial Counselor" },
    { text: "Odio rhoncus ornare ut quam. Molestie vel duis quis scelerisque ut id. In tortor turpis viverra sagittis ultrices nisi.", name: "Ralph Edwards", role: "Math Teacher" },
    { text: "Sagittis nunc egestas leo et malesuada urna risus. Morbi proin et cras aliquam. Diam tellus, amet, hac imperdiet.", name: "Hellena John", role: "Psychology Student" },
    { text: "Sapien, sollicitudin et vitae id et laoreet sapien consectetur. Felis egestas egestas amet aliquam sit euismod.", name: "David Oshodi", role: "Manager" },
    { text: "Sapien, sollicitudin et vitae id et laoreet sapien consectetur. Felis egestas egestas amet aliquam sit euismod.", name: "John Doe", role: "Engineer" }
];

describe('SocialProof Component', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  test('renders static header and cards container', () => {
    renderWithContext(<SocialProof />, mockWindowWidthContext(false));
    expect(screen.getByRole('heading', { name: /What everyone says/i })).toBeInTheDocument();
    expect(screen.getByRole('region', { name: /Social proof/i })).toBeInTheDocument();
    expect(screen.getByTestId('cards-container')).toBeInTheDocument();
  });

  test('renders all user cards with correct content', () => {
    renderWithContext(<SocialProof />, mockWindowWidthContext(false));
    const cardsContainer = screen.getByTestId('cards-container');
    
    componentUserCardData.forEach((userData) => {
      const nameElements = within(cardsContainer).getAllByText(userData.name);
      const cardElement = nameElements.length > 0 ? nameElements[0].closest(`.${"card"}`) : null;

      if (!cardElement) throw new Error(`Card para ${userData.name} não encontrado ou nome não é único no card`);
      
      const escapedText = userData.text.substring(0, 20).replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
      expect(within(cardElement).getByText(new RegExp(escapedText, 'i'))).toBeInTheDocument();
      expect(within(cardElement).getByText(userData.role)).toBeInTheDocument();
      const userImage = within(cardElement).getByAltText(`imagem de ${userData.name}`);
      expect(userImage).toBeInTheDocument();
      expect(userImage).toHaveAttribute('src', '/img.jpg');
    });
  });


  describe('Desktop View Specifics', () => {
    test('renders prev/next buttons and handles scroll on click', () => {
      renderWithContext(<SocialProof />, mockWindowWidthContext(false));
      const cardsContainer = screen.getByTestId('cards-container');
      cardsContainer.scrollBy = jest.fn();

      const prevButton = screen.getByAltText('prev');
      expect(prevButton).toBeInTheDocument();
      fireEvent.click(prevButton);
      expect(cardsContainer.scrollBy).toHaveBeenCalledWith({ left: -300, behavior: "smooth" });

      const nextButton = screen.getByAltText('next');
      expect(nextButton).toBeInTheDocument();
      fireEvent.click(nextButton);
      expect(cardsContainer.scrollBy).toHaveBeenCalledWith({ left: 300, behavior: "smooth" });
      expect(cardsContainer.scrollBy).toHaveBeenCalledTimes(2);
    });

    test('drag to scroll functionality attempts to set scrollLeft', () => {
      renderWithContext(<SocialProof />, mockWindowWidthContext(false));
      const cardsContainer = screen.getByTestId('cards-container');

      Object.defineProperty(cardsContainer, 'offsetLeft', {
        configurable: true,
        value: 50,
      });
      
      cardsContainer.scrollLeft = 100; 
      
      fireEvent.mouseDown(cardsContainer, { pageX: 100 });
      fireEvent.mouseMove(cardsContainer, { pageX: 80 });
      cardsContainer.scrollLeft = 124; 
      
      expect(cardsContainer.scrollLeft).toBe(124);
      
      fireEvent.mouseUp(cardsContainer);
      fireEvent.mouseLeave(cardsContainer);
    });
  });

  describe('Mobile View Specifics', () => {
    test('does not render prev/next buttons', () => {
      renderWithContext(<SocialProof />, mockWindowWidthContext(true));
      expect(screen.queryByAltText('prev')).not.toBeInTheDocument();
      expect(screen.queryByAltText('next')).not.toBeInTheDocument();
    });
  });

  test('matches snapshot on desktop', () => {
    const { container } = renderWithContext(<SocialProof />, mockWindowWidthContext(false));
    expect(container.firstChild).toMatchSnapshot();
  });

  test('matches snapshot on mobile', () => {
    const { container } = renderWithContext(<SocialProof />, mockWindowWidthContext(true));
    expect(container.firstChild).toMatchSnapshot();
  });
});