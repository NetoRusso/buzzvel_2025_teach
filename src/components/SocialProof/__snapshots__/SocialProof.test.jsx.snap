// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialProof Component matches snapshot on desktop 1`] = `
<section
  aria-label="Social proof"
  class="container"
>
  <div
    class="content"
  >
    <div
      class="header"
    >
      <h2
        class="header_title"
      >
        What everyone says
      </h2>
      <div
        class="header_btns"
      >
        <img
          alt="prev"
          class="header_btn"
          height="48"
          loading="lazy"
          src="/img.jpg"
          width="48"
        />
        <img
          alt="next"
          class="header_btn"
          height="48"
          loading="lazy"
          src="/img.jpg"
          width="48"
        />
      </div>
    </div>
    <div
      class="cards_box"
      data-testid="cards-container"
    >
      <div
        class="card"
      >
        <p
          class="card_text"
        >
          Lacus vestibulum ultricies mi risus, duis non, volutpat nullam non. Magna congue nisi maecenas elit aliquet eu sed consectetur.
        </p>
        <div
          class="card_ass"
        >
          <img
            alt="imagem de Hellen Jummy"
            class="card_img"
            height="64"
            loading="lazy"
            src="/img.jpg"
            width="64"
          />
          <div
            class="card_info"
          >
            <p
              class="card_name"
            >
              Hellen Jummy
            </p>
            <p
              class="card_role"
            >
              Financial Counselor
            </p>
          </div>
        </div>
      </div>
      <div
        class="card"
      >
        <p
          class="card_text"
        >
          Odio rhoncus ornare ut quam. Molestie vel duis quis scelerisque ut id. In tortor turpis viverra sagittis ultrices nisi.
        </p>
        <div
          class="card_ass"
        >
          <img
            alt="imagem de Ralph Edwards"
            class="card_img"
            height="64"
            loading="lazy"
            src="/img.jpg"
            width="64"
          />
          <div
            class="card_info"
          >
            <p
              class="card_name"
            >
              Ralph Edwards
            </p>
            <p
              class="card_role"
            >
              Math Teacher
            </p>
          </div>
        </div>
      </div>
      <div
        class="card"
      >
        <p
          class="card_text"
        >
          Sagittis nunc egestas leo et malesuada urna risus. Morbi proin et cras aliquam. Diam tellus, amet, hac imperdiet.
        </p>
        <div
          class="card_ass"
        >
          <img
            alt="imagem de Hellena John"
            class="card_img"
            height="64"
            loading="lazy"
            src="/img.jpg"
            width="64"
          />
          <div
            class="card_info"
          >
            <p
              class="card_name"
            >
              Hellena John
            </p>
            <p
              class="card_role"
            >
              Psychology Student
            </p>
          </div>
        </div>
      </div>
      <div
        class="card"
      >
        <p
          class="card_text"
        >
          Sapien, sollicitudin et vitae id et laoreet sapien consectetur. Felis egestas egestas amet aliquam sit euismod.
        </p>
        <div
          class="card_ass"
        >
          <img
            alt="imagem de David Oshodi"
            class="card_img"
            height="64"
            loading="lazy"
            src="/img.jpg"
            width="64"
          />
          <div
            class="card_info"
          >
            <p
              class="card_name"
            >
              David Oshodi
            </p>
            <p
              class="card_role"
            >
              Manager
            </p>
          </div>
        </div>
      </div>
      <div
        class="card"
      >
        <p
          class="card_text"
        >
          Sapien, sollicitudin et vitae id et laoreet sapien consectetur. Felis egestas egestas amet aliquam sit euismod.
        </p>
        <div
          class="card_ass"
        >
          <img
            alt="imagem de John Doe"
            class="card_img"
            height="64"
            loading="lazy"
            src="/img.jpg"
            width="64"
          />
          <div
            class="card_info"
          >
            <p
              class="card_name"
            >
              John Doe
            </p>
            <p
              class="card_role"
            >
              Engineer
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`SocialProof Component matches snapshot on mobile 1`] = `
<section
  aria-label="Social proof"
  class="container"
>
  <div
    class="content"
  >
    <div
      class="header"
    >
      <h2
        class="header_title"
      >
        What everyone says
      </h2>
    </div>
    <div
      class="cards_box"
      data-testid="cards-container"
    >
      <div
        class="card"
      >
        <p
          class="card_text"
        >
          Lacus vestibulum ultricies mi risus, duis non, volutpat nullam non. Magna congue nisi maecenas elit aliquet eu sed consectetur.
        </p>
        <div
          class="card_ass"
        >
          <img
            alt="imagem de Hellen Jummy"
            class="card_img"
            height="64"
            loading="lazy"
            src="/img.jpg"
            width="64"
          />
          <div
            class="card_info"
          >
            <p
              class="card_name"
            >
              Hellen Jummy
            </p>
            <p
              class="card_role"
            >
              Financial Counselor
            </p>
          </div>
        </div>
      </div>
      <div
        class="card"
      >
        <p
          class="card_text"
        >
          Odio rhoncus ornare ut quam. Molestie vel duis quis scelerisque ut id. In tortor turpis viverra sagittis ultrices nisi.
        </p>
        <div
          class="card_ass"
        >
          <img
            alt="imagem de Ralph Edwards"
            class="card_img"
            height="64"
            loading="lazy"
            src="/img.jpg"
            width="64"
          />
          <div
            class="card_info"
          >
            <p
              class="card_name"
            >
              Ralph Edwards
            </p>
            <p
              class="card_role"
            >
              Math Teacher
            </p>
          </div>
        </div>
      </div>
      <div
        class="card"
      >
        <p
          class="card_text"
        >
          Sagittis nunc egestas leo et malesuada urna risus. Morbi proin et cras aliquam. Diam tellus, amet, hac imperdiet.
        </p>
        <div
          class="card_ass"
        >
          <img
            alt="imagem de Hellena John"
            class="card_img"
            height="64"
            loading="lazy"
            src="/img.jpg"
            width="64"
          />
          <div
            class="card_info"
          >
            <p
              class="card_name"
            >
              Hellena John
            </p>
            <p
              class="card_role"
            >
              Psychology Student
            </p>
          </div>
        </div>
      </div>
      <div
        class="card"
      >
        <p
          class="card_text"
        >
          Sapien, sollicitudin et vitae id et laoreet sapien consectetur. Felis egestas egestas amet aliquam sit euismod.
        </p>
        <div
          class="card_ass"
        >
          <img
            alt="imagem de David Oshodi"
            class="card_img"
            height="64"
            loading="lazy"
            src="/img.jpg"
            width="64"
          />
          <div
            class="card_info"
          >
            <p
              class="card_name"
            >
              David Oshodi
            </p>
            <p
              class="card_role"
            >
              Manager
            </p>
          </div>
        </div>
      </div>
      <div
        class="card"
      >
        <p
          class="card_text"
        >
          Sapien, sollicitudin et vitae id et laoreet sapien consectetur. Felis egestas egestas amet aliquam sit euismod.
        </p>
        <div
          class="card_ass"
        >
          <img
            alt="imagem de John Doe"
            class="card_img"
            height="64"
            loading="lazy"
            src="/img.jpg"
            width="64"
          />
          <div
            class="card_info"
          >
            <p
              class="card_name"
            >
              John Doe
            </p>
            <p
              class="card_role"
            >
              Engineer
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
